proxy/HostStatus.cc

```diff
@@ -0,0 +1,89 @@
+/** @file
+
+  Implementation of Host Proxy routing
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+#include "HostStatus.h"
+
+HostStatus *HostStatus::p_instance = nullptr;
+
+HostStatus::HostStatus() : hosts_statuses(ink_hash_table_create(InkHashTableKeyType_String))
+{
+  ink_mutex_init(&hosts_statuses_mutex);
+}
+
+HostStatus::~HostStatus()
+{
+  InkHashTableIteratorState ht_iter;
+  InkHashTableEntry *ht_entry = nullptr;
+  ht_entry                    = ink_hash_table_iterator_first(hosts_statuses, &ht_iter);
+
+  while (ht_entry != nullptr) {
+    char *value = static_cast<char *>(ink_hash_table_entry_value(hosts_statuses, ht_entry));
+    ats_free(value);
+    ht_entry = ink_hash_table_iterator_next(hosts_statuses, &ht_iter);
+  }
+  ink_hash_table_destroy(hosts_statuses);
+
+  ink_mutex_destroy(&hosts_statuses_mutex);
+}
+
+HostStatus *
+HostStatus::instance()
+{
+  return p_instance;
+}
+
+HostStatus *
+HostStatus::init()
+{
+  if (!p_instance) {
+    p_instance = new HostStatus();
+  }
+  return p_instance;
+}
+
+void
+HostStatus::setHostStatus(const char *key, const HostStatus_t status)
+{
+  HostStatus_t *_status;
+
+  ink_mutex_acquire(&hosts_statuses_mutex);
+  if (ink_hash_table_lookup(hosts_statuses, key, (void **)&_status)) {
+    *_status = status;
+  } else {
+    Debug("hosts_statuses", "In HostConfigParams::setHostStatus(): key: %s, status: %d", key, status);
+    _status  = static_cast<HostStatus_t *>(ats_malloc(sizeof(HostStatus_t)));
+    *_status = status;
+  }
+  ink_hash_table_insert(hosts_statuses, key, _status);
```

Comment:
Do you always want to insert this, even if it was found?

Reply:
ink_hash_table_insert() will also replace any previous value if the entry exists.

Reply:
@amc I'm fixing this.  I do need to allocate a HostStatus_t though as the has table binds a pointer to the value.  The HostStatus destructor frees the values allocated.